### ISSUE TITLE
Discrepency between tag counts in search results and tag page's count

### ISSUE THREAD

**Issue by jeremyf** [forem/forem#16742]
The "knowledge" tag at https://dev.to/tags?q=knowledge says that there are 85 published posts.  When I go to https://dev.to/t/knowledge, the number of posts says 34.

Similarly "knowledgebase" https://dev.to/tags?q=knowledge says says 512 posts, but https://dev.to/t/knowledgebase says 5 posts.

I would expect these to have at least close to the same count.

**Comment by jeremyf** [comment]
For the individual tag, we use the following logic:

https://github.com/forem/forem/blob/dce7c4b2752f154ff6518efa2be0cec6c603f370/app/controllers/stories/tagged_articles_controller.rb#L57-L59

For the search page, we use the `Tag#taggings_count`

https://github.com/forem/forem/blob/dce7c4b2752f154ff6518efa2be0cec6c603f370/app/views/tags/index.html.erb#L58-L63

These are very different logic assumptions.

**Comment by amywtlin** [comment]
Yes. We should use the number of published & approved posts on both the search result page & the tag listing page.

**Comment by thejwuscript** [comment]
I would like to work on this issue.

**Comment by blaqturtle** [comment]
It seems the issue is to do with the two pages (filtered and unfiltered tags) being cached at different times. Since the two versions of the page are read with different routes, `/tags` for unfiltered vs `/tags?q=:tagname` for filtered, they are cached separately under different keys.

Since the caches are separate and happen at different times, the actual count of articles associated with a tag may be different when they are cached. Add to that the fact that the `/tags` page cache (whether filtered or not) expires after 24 hours and you end up with two pages cached with different values and expiring at different times. When the caches do expire, likely at different times, they are updated with new values that will likely be different from each other.

My proposed solution is to cache the article count for each tag independently, with an expiry window longer than the page cache (24 hours). Both filtered and unfiltered pages would read from this shared count cache rather than computing the count on render. This ensures both pages reflect the same value regardless of when their respective page caches were created or expire. The implementation will also consider what events should invalidate the count cache — tag updates, significant article state changes — so that the cached count doesn't become unreasonably stale.

One implementation approach is to use the Rails fragment cache for each tag card, caching the entire rendered card so it appears identically on both filtered and unfiltered pages. The cache key would include values that trigger re-rendering when the tag's name, description, or other attributes change.

The expiry window for the shared cache will need to account for the fact that page caches may not be refreshed immediately on expiry, extending the effective window unpredictably. This will require careful consideration during implementation.
